### PR TITLE
Add Ability to Override Service Names

### PR DESCRIPTION
**Pull request by blakeromano**
Resolves https://github.com/open-telemetry/opentelemetry-operator/issues/938

### CODE REVIEW

**Comment by blakeromano** [comment]
I just added some documentation in the readme about this 👍  @mat-rumian

File: pkg/instrumentation/sdk.go [hunk 1]
```diff
@@ -208,6 +208,9 @@ func (i *sdkInjector) injectCommonSDKConfig(ctx context.Context, otelinst v1alph
 }
 
 func chooseServiceName(pod corev1.Pod, resources map[string]string, index int) string {
+	if name := strings.Split(pod.Annotations["instrumentation.opentelemetry.io/service-name"], ",")[index]; name != "" {
```
Comment (pavolloffay) [inline pkg/instrumentation/sdk.go:211]: please define this as a constant

File: README.md [hunk 2]
```diff
@@ -254,6 +254,18 @@ spec:
 
 In the above case, `myapp` and `myapp2` containers will be instrumented, `myapp3` will not.
 
+#### Override Service Name
+
+The service name for containers goes in the following priority list
```
Comment (pavolloffay) [inline README.md:259]: missing `:` at the end

File: README.md [hunk 3]
```diff
@@ -254,6 +254,18 @@ spec:
...
+
+* The `instrumentation.opentelemetry.io/service-name` pod annotation (can be `,` seperated list for multiple container services)
```
Comment (pavolloffay) [inline README.md:261]: The order does not look right. Some of the object types have the same priority - deployment, SS, CJ, Job

**Review by pavolloffay — COMMENTED** [review]
The feature makes sens, but I would like to see a e2e test for this capability

File: README.md (hunk 2 above)
Comment (pavolloffay) [reply README.md:259]: We should provide some more information about how the service name is determined e.g.: 

> The operator automatically determines the most suitable service name from the Kubernetes object names. However the service name can be overridden `TBD` annotation in the pod specification. Follows the precedence order for determining the service name:

File: pkg/instrumentation/sdk.go [hunk 4]
```diff
@@ -208,6 +208,9 @@ func (i *sdkInjector) injectCommonSDKConfig(ctx context.Context, otelinst v1alph
 }
 
 func chooseServiceName(pod corev1.Pod, resources map[string]string, index int) string {
```
Comment (pavolloffay) [inline pkg/instrumentation/sdk.go:210]: please add unit test as well

**Comment by blakeromano** [comment]
@pavolloffay I made the changes you suggested. I am not super familiar with kuttl but if you feel like e2e tests are needed for this I can do my best to write one up.

File: README.md [hunk 5]
```diff
@@ -254,6 +254,16 @@ spec:
 
 In the above case, `myapp` and `myapp2` containers will be instrumented, `myapp3` will not.
 
+#### Override Service Name
+
+The operator automatically determines the most suitable service name from the Kubernetes object names. However the service name can be overridden by the `instrumentation.opentelemetry.io/service-name` annotation in the pod specification. Follows the precedence order for determining the service name:
+
+1. The `instrumentation.opentelemetry.io/service-name` pod annotation (can be `,` seperated list for multiple container services)
+2. The Deployment Name
```
Comment (pavolloffay) [inline README.md:262]: iirc the deployment name should be on the same line as ss, cj and job

File: pkg/instrumentation/sdk.go [hunk 6]
```diff
@@ -208,6 +208,10 @@ func (i *sdkInjector) injectCommonSDKConfig(ctx context.Context, otelinst v1alph
...
+	serviceNames := strings.Split(pod.Annotations["instrumentation.opentelemetry.io/service-name"], ",")
```
Comment (pavolloffay) [inline pkg/instrumentation/sdk.go:211]: extract the annotation to a constant

Comment (blakeromano) [reply pkg/instrumentation/sdk.go:211]: I did the split in the case of people with multiple containers. for you to be able to override the different container names

File: tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+apiVersion: opentelemetry.io/v1alpha1
+kind: Instrumentation
+metadata:
+  name: java
+spec:
+  env:
```
Comment (pavolloffay) [inline tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml:6]: somebody can define OTEL_SERVICE_NAME here. What is the precedence order of setting the service name? We should document this in the readme and have a test.

**Review by pavolloffay — COMMENTED** [review]
LGTM, just one comment on a corner case that I didn't catch before.

File: tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml (hunk 7 above)
Comment (blakeromano) [reply tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml:6]: I honestly am not crazy sure on these answers. If you don't mind helping me fill the gaps.

Comment (yuriolisa) [reply tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml:6]: @blakeromano, I'm going to check what is missing to help you out.

Comment (blakeromano) [reply tests/e2e/instrumentation-service-name-annotation/00-install-instrumentation.yaml:6]: @yuriolisa I apologize been busy I was wondering if you had time to help me out 😸

**Comment by jaronoff97** [comment]
@blakeromano i think this is going to be superseded by #2330 are you interested in continuing this work?

**Comment by atoulme** [comment]
Closing as inactive. Please comment if you would like to reopen this pull request.
